### DISCUSSION TITLE
マルチプレイSDKで複数のアバターをロードしたいです。

### DISCUSSION THREAD

**Discussion by SK0519** [pixiv/vroid-sdk-developers#163] [Q&A]
現在Photon Fusion 2を利用して各々のアバターで遊べるマルチプレイのゲームを開発中です。
Fusion2のネットワークプロパティを利用してダウンロードライセンスIDを取得後に、自動でお互いのダウンロードライセンスIDを交換できるシステムを実装しました。

マルチプレイExampleの「DownloadLicenseIdField」をいくつか複製し、ダウンロードライセンスIDを複製した「DownloadLicenseIdField」に転写できるようになっています(開発段階に視認できるようにするため。)
同じようにExampleの「Routes」と「CharactersRoot」を複製し、それぞれに対応するものをアタッチしています。
(例：「Routes0」のEventHandlerには「CharactersRoot0」、「CharactersRoot1」のDownloadLicenseIdFieldには「DownloadLicenseIdField1」といった感じです)

スクリーンショットの「アバターをダウンロード」ボタンをクリックした際にこれらのボタンクリックイベントを動かして同時に複数のアバターをダウンロード・表示できるように実装したつもりなのですが、ユーザー１はアバター１、ユーザー２はアバター１、ユーザー３はアバター２をダウンロード・表示するような挙動になってしまっています。
３日ほどExampleのコードと向き合ったのですが自分の能力では原因を特定することができませんでした。
複数のInputFieldを参照して同時に複数のアバターをダウンロード・表示させる方法はございますでしょうか？
お手数おかけしてすみませんがよろしくお願いします。

![スクリーンショット 2024-06-28 201432](https://github.com/pixiv/vroid-sdk-developers/assets/174109428/9a1db13a-0425-4fa9-91f6-588a60e80524)

**Comment by kajiina** [comment]
ユーザーとアバターがずれている原因については、 `MultiplayModelLoader.LoadVrm` を呼び出す際の `licenseId` がずれてしまっているためだと考えられます。

他のユーザーのモデルを表示できていることから、ダウンロードライセンスIDを他のユーザーに送信することはできていると思いますので、ダウンロードライセンスIDを他のユーザーから受信した際の処理にあると思われる `MultiplayModelLoader.LoadVrm` の `licenseId` を確認してみてください

**Reply by SK0519** [reply]
お返事ありがとうございます。
「アバターをダウンロード」ボタンには複数の「Routes」を設定してそれぞれに「UseDownloadLicense」を実行させているのですが、反応がありません。コードをたどって「OnUpdateAsync」の同時に実行されないようにするフラグをコメントアウトしたのですが結果は同じでした。
他に同時にダウンロードを走らせないようにする部分やVroid HUBへの接続にストップをかけるような部分などはありますでしょうか？

**Reply by SK0519** [reply]
「ApiController.LoginToVroidHub」の「GetAccountInfo」でアカウント情報を取得に失敗しているようです。
「GetAccountInfo」内のif (_model.CurrentUser != null)｛～～～でCueewntUserがnullになっていることが原因でした。
もともとの「ダウンロードライセンスを利用」ボタンは問題なく動作するため比較しているのですが原因が特定できません。

**Reply by kajiina** [reply]
返信遅くなり申し訳ありません。
CurrentUserがnilになっている件について、予想ですがPhotonで接続している別ユーザーがVRoid Hubにログインできていないためだと思います。マルチプレイで他ユーザーのモデルをダウンロードする際でも、自身のVRoid Hubアカウントにログインする必要があります

**Reply by SK0519** [reply]
テストのため複数台のPC、複数のアカウントで実験を行っているのですが、ログイン状態が確認できているにもかかわらず「接続失敗」が表示されてしまいます。また、自分のダウンロードライセンスIDを利用する「Download」ボタンを押した場合でも「接続失敗」と表示されてしまいます。デバッグモードにしてもログインに関わる部分はこちらには見えないため挙動を知ることができません。
現在、Exampleシーンに複数のアバターを表示してそれをゲームシーンに持ち込む方向で考えているのですが、難しいようであれば他のアプローチをとった方がよいのでしょうか？
![image](https://github.com/pixiv/vroid-sdk-developers/assets/174109428/bc6e0d64-7a94-4791-8025-a0c4921a85f1)

**Reply by kajiina** [reply]
申し訳ございません、返信遅くなりました
すみませんExampleシーンを使う場合についてですが、Routesそのものはコピーして動くようにはなっていないようなので、Routesは以下のCharacterRootの指定を配列に変更する方法でこちらでも試してみます
